src/libappimage/core/impl/TraversalType2.cpp

```diff
@@ -0,0 +1,296 @@
+            extractFile(inode, target);
+            break;
+        case SQUASHFS_SYMLINK_TYPE:
+        case SQUASHFS_LSYMLINK_TYPE:
+            extractSymlink(inode, target);
+            break;
+        default:
+            throw AppImageError("AppImage Type 2 inode.base.inode_type " +
+                                std::to_string(inode.base.inode_type) + " not supported yet.");
+    }
+}
+
+sqfs_err TraversalType2::sqfs_stat(sqfs* fs, sqfs_inode* inode, struct stat* st) {
+    // code borrowed from the "extract.c" file at squashfuse
+    sqfs_err err;
+    uid_t id;
+
+    memset(st, 0, sizeof(*st));
+    // fill stats
+    st->st_mode = inode->base.mode;
+    st->st_nlink = inode->nlink;
+    st->st_mtime = st->st_ctime = st->st_atime = inode->base.mtime;
+
+    if (S_ISREG(st->st_mode)) {
+        /* FIXME: do symlinks, dirs, etc have a size? */
+        st->st_size = inode->xtra.reg.file_size;
+        st->st_blocks = st->st_size / 512;
+    } else if (S_ISBLK(st->st_mode) || S_ISCHR(st->st_mode)) {
+        st->st_rdev = sqfs_makedev(inode->xtra.dev.major,
+                                   inode->xtra.dev.minor);
+    } else if (S_ISLNK(st->st_mode)) {
+        st->st_size = inode->xtra.symlink_size;
+    }
+
+    st->st_blksize = fs->sb.block_size; /* seriously? */
+
+    err = sqfs_id_get(fs, inode->base.uid, &id);
+    if (err)
+        return err;
+    st->st_uid = id;
+    err = sqfs_id_get(fs, inode->base.guid, &id);
+    st->st_gid = id;
+    if (err)
+        return err;
+
+    return SQFS_OK;
+}
+
+void TraversalType2::extractDir(const std::string& target) {
+    if (access(target.c_str(), F_OK) == -1) { // The directory doesn't exists
+        if (mkdir(target.c_str(), 0755) == -1) // Create new directory with 755 permissions
+            throw AppImageError("mkdir error at " + target);
+    }
+}
+
+void TraversalType2::extractFile(sqfs_inode inode, const std::string& target) {
+    // Read inode stats
+    struct stat st = {};
+    if (sqfs_stat(&fs, &inode, &st) != 0)
+        throw AppImageReadError("sqfs_stat error");
+
+    // open read stream
+    auto& istream = read();
+
+    // open write stream
+    ofstream targetFile(target);
+
+    // trasnfer data
+    targetFile << istream.rdbuf();
+    targetFile.close();
+
+    // set file stats
+    chmod(target.c_str(), st.st_mode);
+}
+
+void TraversalType2::extractSymlink(sqfs_inode inode, const std::string& target) {
+    // read the target link path size
+    size_t size;
+    sqfs_readlink(&fs, &inode, nullptr, &size);
+
+    char buf[size];
+
+    // read the target link in buf
+    int ret = sqfs_readlink(&fs, &inode, buf, &size);
+    if (ret != 0)
+        throw AppImageReadError("sqfs_readlink error");
+
+    // remove any existent link at t
+    unlink(target.c_str());
+
+    ret = symlink(buf, target.c_str());
+    if (ret != 0)
+        throw AppImageReadError("symlink error at " + target);
+}
+
+istream& TraversalType2::read() {
```

Comment:
The documentation is on the base class method. A reference is returned because the Traversal totally owns and controls this object and it "becomes invalid every time next is called". We cannot give away this object.

Reply:
Seems valid. A small comment will help future readers who need to revise this code, though.

Reply:
Docstrings added in the base class, as it applies to all derivatives.